// Copyright 2025 The Protoscribe Authors.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

// In morphophonology the prosodic templates are morphological constructions
// that directly constrain the phonological shape of the derived stem.

syntax = "proto2";

package protoscribe;

message ProsodyTemplate {
  // Maximum number of distinct most segments to sample from the phonological
  // inventory when contructing syllables.
  optional int32 number_of_phonemes = 1;

  // Proportion of the above selected segments to actually use.
  optional float probability_of_selection = 2 [default = 1.0];

  // List of syllabic templates such as ["CV", "CVC", ...].
  repeated string list_of_templates = 3;

  // Maximum number of syllables to allow.
  optional int32 max_syllables = 4 [default = 1];

  // Minimum number of syllables to allow.
  optional int32 min_syllables = 5 [default = 1];

  // Add in sesquisyllables, where a sesquisyllable is defined as prosodic word
  // consisting of an unstressed "minor" syllable followed by a stressed "major"
  // syllable.
  optional bool use_sesquisyllabic = 6 [default = false];

  // A list of ISO-639-2 language codes from which to select phonemes.
  repeated string languages = 7;
}
